File: src/ollama_bench/benchmarks/generation.py
from __future__ import annotations

from ollama_bench.benchmarks.base import BenchmarkContext, BenchmarkResultRecord
from ollama_bench.benchmarks.metrics import generation_metrics, preview
from ollama_bench.ollama.client import OllamaClient
from ollama_bench.prompts import SHORT_PROMPT


class BasicGenerationLatencyBenchmark:
    id = "generation-latency"
    name = "Basic generation latency"
    description = "Measures wall time, time to first token, Ollama duration fields, and output token throughput."

    def run(self, client: OllamaClient, model: str, context: BenchmarkContext) -> list[BenchmarkResultRecord]:
        records: list[BenchmarkResultRecord] = []
        options = context.generation_options(temperature=0.0)
        for run_index in range(1, context.runs + 1):
            try:
                result = client.generate_stream(
                    model,
                    SHORT_PROMPT,
                    options=options,
                    keep_alive=context.keep_alive,
                    timeout=context.timeout,
                )
                records.append(
                    BenchmarkResultRecord(
                        benchmark_id=self.id,
                        benchmark_name=self.name,
                        model=model,
                        run_index=run_index,
                        prompt_name="short",
                        metrics=generation_metrics(result),
                        response_preview=preview(result.response),
                        metadata={"options": options},
                    )
                )
            except Exception as exc:  # noqa: BLE001, converted to result so multi-model runs continue
                records.append(
                    BenchmarkResultRecord(
                        benchmark_id=self.id,
                        benchmark_name=self.name,
                        model=model,
                        run_index=run_index,
                        prompt_name="short",
                        metrics={},
                        error=str(exc),
                        metadata={"options": options},
                    )
                )
        return records

File: src/benchmarks/generation.rs
use std::collections::HashMap;

use crate::benchmarks::base::{
    Benchmark, BenchmarkContext, BenchmarkProgressSink, BenchmarkResultRecord, BenchmarkStepStatus,
    BenchmarkStepUpdate,
};
use crate::benchmarks::metrics::{generation_metrics, preview};
use crate::benchmarks::registry::BenchmarkSuite;
use crate::prompts::SHORT_PROMPT;
use crate::providers::ProviderClient;
use crate::utils::error_chain;

pub struct BasicGenerationLatencyBenchmark;

impl Benchmark for BasicGenerationLatencyBenchmark {
    fn id(&self) -> &str {
        "chat-generation"
    }

    fn name(&self) -> &str {
        "Basic generation latency"
    }

    fn description(&self) -> &str {
        "Measures wall time, time to first token, usage fields, and output token throughput."
    }

    fn suite(&self) -> BenchmarkSuite {
        BenchmarkSuite::Llm
    }

    fn planned_steps(&self, context: &BenchmarkContext) -> u32 {
        context.runs
    }

    fn run(
        &self,
        client: &ProviderClient,
        model: &str,
        context: &BenchmarkContext,
        progress: &mut dyn BenchmarkProgressSink,
    ) -> Vec<BenchmarkResultRecord> {
        let mut records = Vec::new();
        let options = context.request_options(Some(0.0), None);
        let messages = serde_json::json!([
            {"role": "user", "content": SHORT_PROMPT}
        ]);
        let total_steps = self.planned_steps(context);

        for run_index in 1..=context.runs {
            progress.on_step(BenchmarkStepUpdate {
                status: BenchmarkStepStatus::Started,
                step_index: run_index,
                total_steps,
                run_index: Some(run_index),
                prompt_name: Some("short".to_string()),
                message: "Issuing chat completion request".to_string(),
            });
            let record = match client.chat_completion(
                model,
                messages.clone(),
                context.max_tokens,
                0.0,
                true,
                Some(&options),
            ) {
                Ok(result) => BenchmarkResultRecord {
                    benchmark_id: self.id().to_string(),
                    benchmark_name: self.name().to_string(),
                    model: model.to_string(),
                    run_index: Some(run_index),
                    prompt_name: Some("short".to_string()),
                    metrics: generation_metrics(&result),
                    response_preview: Some(preview(&result.response_text, 180)),
                    error: None,
                    metadata: Some({
                        let mut m = HashMap::new();
                        m.insert("options".to_string(), options.clone());
                        m
                    }),
                },
                Err(e) => BenchmarkResultRecord {
                    benchmark_id: self.id().to_string(),
                    benchmark_name: self.name().to_string(),
                    model: model.to_string(),
                    run_index: Some(run_index),
                    prompt_name: Some("short".to_string()),
                    metrics: HashMap::new(),
                    response_preview: None,
                    error: Some(error_chain(&*e)),
                    metadata: Some({
                        let mut m = HashMap::new();
                        m.insert("options".to_string(), options.clone());
                        m
                    }),
                },
            };
            records.push(record);
            progress.on_step(BenchmarkStepUpdate {
                status: BenchmarkStepStatus::Completed,
                step_index: run_index,
                total_steps,
                run_index: Some(run_index),
                prompt_name: Some("short".to_string()),
                message: "Completed chat completion request".to_string(),
            });
        }
        records
    }
}
